backend/resources/org/akvo/dash/migrations_tenants/007-database-functions.up.sql

```diff
@@ -39,50 +26,169 @@ BEGIN
   END IF;
   RETURN '"' || tmp || '"';
 END;
-$BODY$
+$$
   LANGUAGE plpgsql IMMUTABLE STRICT;
 
-CREATE OR REPLACE FUNCTION to_date(val jsonb, format text, raise_on_error boolean)
+CREATE OR REPLACE FUNCTION lumen_to_date(val jsonb, default_val jsonb,
+  parse_format text, raise_on_error boolean)
   RETURNS jsonb AS
-$BODY$
+$$
 DECLARE
-  tmp text = trim(both '"' from val::text);
+  tval text = val::text;
+  tmp text = trim(both '"' from tval);
+  t1 text;
 BEGIN
-  RETURN date_part('epoch', to_timestamp(tmp, format))::numeric;
+  IF tmp = tval THEN
+    RAISE EXCEPTION 'Value was not text: %', tval;
+  END IF;
+  RETURN date_part('epoch', to_timestamp(tmp, parse_format))::numeric;
 EXCEPTION
   WHEN OTHERS THEN
-  IF raise_on_error THEN
-    RAISE EXCEPTION 'Unable to convert % to date value', val;
-  ELSE
-    RETURN NULL;
-  END IF;
+    GET STACKED DIAGNOSTICS t1 = MESSAGE_TEXT;
+    IF raise_on_error THEN
+      RAISE EXCEPTION 'Unable to convert % to date value - msg: %', val, t1;
+    ELSE
+      RETURN default_val;
+    END IF;
 END;
-$BODY$
+$$
   LANGUAGE plpgsql IMMUTABLE STRICT;
 
-CREATE OR REPLACE FUNCTION to_date(val jsonb, format text)
+CREATE OR REPLACE FUNCTION lumen_titlecase(val jsonb)
   RETURNS jsonb AS
-$BODY$
+$$
 BEGIN
-  RETURN to_date(val, format, false);
+  RETURN initcap(val::text);
 END;
-$BODY$
+$$
   LANGUAGE plpgsql IMMUTABLE STRICT;
 
-CREATE OR REPLACE FUNCTION titlecase(val jsonb)
+CREATE OR REPLACE FUNCTION lumen_trim_double(val jsonb)
   RETURNS jsonb AS
-$BODY$
+$$
 BEGIN
-  RETURN initcap(val::text);
+  RETURN regexp_replace(val::text, '\s+', ' ', 'g');
```

Comment:
Should `lumen_trim_double` collapse newlines and tabs into a single space as well?